core(inspector-issues): collect all issue types

adamread opened the pull request:
The inspector issues gatherer is now collecting all of the issues that will be reported via the devtools protocol.
The tests directly related to the issues gatherer have been updated, but there are still some other tests only tangentally related that are now broken.

<!--
Thank you for submitting a pull request!
See CONTRIBUTING.MD for help in getting a change landed.
  https://github.com/GoogleChrome/lighthouse/blob/master/CONTRIBUTING.md
-->

**Summary**
<!-- What kind of change does this PR introduce? -->
<!-- Is this a bugfix, feature, refactoring, build related change, etc? -->
This is an update to the inspector issues gatherer and audit to cover all of the issues currently reported via the devtools protocol and shown in the issues panel. Part of the goal of this change is to update the gatherer and audit to require less maintenance for future updates.

<!-- Describe the need for this change -->
A bunch of issues have been added to the issues panel since the original lighthouse implementation.

<!-- Link any documentation or information that would help understand this change -->

**Related Issues/PRs**
<!-- Provide any additional information we might need to understand the pull request -->
Issue #13147

adamread commented:
This change has broken a bunch of other tests in Sample JSON, Security, and Fraggle Rock. It would be good to get some feedback before I go and update all the tests too.

connorjclark commented on lighthouse-core/gather/gatherers/inspector-issues.js:
```diff
@@ -59,57 +59,60 @@ class InspectorIssues extends FRGatherer {
+    /** @type {Array<keyof LH.Artifacts['InspectorIssues']>} */
+    const keys = Object.keys(artifact);
+    for (const key of keys) {
+      /** @type {`${key}Details` | `wasmCrossOriginModuleSharingIssue`} */
```
cc @brendankenny 

![image](https://user-images.githubusercontent.com/4071474/145310341-b4146612-685a-4c30-b820-715db1fab2c3.png)


> A 'const' assertions can only be applied to references to enum members, or string, number, boolean, array, or object literals.ts(1355)

too bad

adamread replied:
Just making sure there was nothing you wanted here before I resolve the comment.

adamread commented on lighthouse-core/gather/gatherers/inspector-issues.js:
```diff
@@ -59,57 +59,60 @@ class InspectorIssues extends FRGatherer {
+      const allDetails = this._issues.map(issue => issue.details[detailsKey]);
+      for (const detail of allDetails) {
+        if (detail) {
+          if (Object.prototype.hasOwnProperty.call(detail, 'request')) {
```
Thanks. I'm still figuring out how to work with TS when things like this come up.

adamread commented on lighthouse-core/gather/gatherers/inspector-issues.js:
```diff
@@ -59,57 +59,60 @@ class InspectorIssues extends FRGatherer {
+      const detailsKey = (key === 'wasmCrossOriginModuleSharingIssue' ? `${key}` : `${key}Details`);
+      const allDetails = this._issues.map(issue => issue.details[detailsKey]);
+      for (const detail of allDetails) {
+        if (detail) {
```
I'm not sure details will ever be missing in real life. It is in some of the existing tests though, so I added the check here. It seems like the test here offers some safety, even if this shouldn't happen when the issues are coming from devtools.

If this should never happen, should I maybe update the check to fail harder?

connorjclark replied:
I overlooked that these are defined as optional in the interface, so they check does make sense, at least at the type-level. It ought to always exist but the types are the types and we must abide :)

no need to do anything special, let's just do the early-continue style instead

adamread commented on types/artifacts.d.ts:
```diff
@@ -553,12 +553,21 @@ declare module Artifacts {
+    wasmCrossOriginModuleSharingIssue: LH.Crdp.Audits.WasmCrossOriginModuleSharingIssueDetails[];
+    genericIssue: LH.Crdp.Audits.GenericIssueDetails[];
+    deprecationIssue: LH.Crdp.Audits.DeprecationIssueDetails[];
   }
```
This is the order they're defined in the gatherer, which is the order they're defined in in the devtools protocol InspectorIssueDetails interface. I'm not sure why they're not sorted in the original interface. Does that make a difference?

connorjclark replied:
We have the liberty to resort our own usage of these fields (so I'd also suggest sorting here and elsewhere we define this artifact type). I don't think the ordering in the third-party CRDP type definitions being different will cause any contention.

So it just comes down to what is slightly simpler when looking at the raw artifact JSON, and ordered keys are good for debugging.

adamread commented on types/artifacts.d.ts:
```diff
@@ -553,12 +553,21 @@ declare module Artifacts {
   }
 
   interface InspectorIssues {
```
Of course. When I saw that this change to the gatherer breaks a bunch of other stuff, I thought it might be better to break this into it's own PR. 

Would it be better to update this PR with all of the work?

connorjclark replied:
Keeping it separate is the preference. I just wanted to write down the new items

adamraine commented:
Looks like `clientHintIssueDetails` is a new one now. Caught by our [notifier](https://github.com/GoogleChrome/lighthouse/blob/master/third-party/chromium-synchronization/inspector-issueAdded-types-test.js) in https://github.com/GoogleChrome/lighthouse/pull/13502.

adamread commented:
Updated the gatherer and updated the version of devtools-protocol to include clientHintIssueDetails and updated everything else to fix the breaking tests.

@connorjclark - Do the sample artifacts get checked in too?

connorjclark commented:
Yes, but we have a command to isolate changes to the sample artifacts to just a single artifact. try this:

```
# in case you have local changes
git checkout origin/master -- lighthouse-core/test/results/artifacts

yarn update:sample-artifacts InspectorIssues
```

connorjclark reviewed (approved):
LGTM! It'll be nice to finally get back up-to-date with these.
